misc/config_tools/launch_config/launch_cfg_gen.py

```diff
@@ -1,269 +1,364 @@
+        if not m:
+            continue
+        script.add_passthru_device(int(m.group(1), 16), int(m.group(2), 16), int(m.group(3), 16), options="igd-vf")
+
+    ###
+    # Miscellaneous
+    ###
+    script.add_dynamic_dm_parameter("add_rtvm_options")
+    script.add_dynamic_dm_parameter("add_logger_settings", "console=4 kmsg=3 disk=5")
+
+    ###
+    # Lastly, conclude the device model parameters with the VM name
+    ###
+    script.add_plain_dm_parameter(f"{vm_name}")
+
+    return script
+
+def main(board_xml, scenario_xml, launch_xml, user_vmid, out_dir):
+    board_etree = etree.parse(board_xml)
+    scenario_etree = etree.parse(scenario_xml)
+    launch_etree = etree.parse(launch_xml)
+
+    sos_vm_id = eval_xpath(scenario_etree, "//vm[vm_type='SERVICE_VM']/@id")
+    post_vms = scenario_etree.xpath("//vm[starts-with(vm_type, 'POST_')]")
+    if sos_vm_id is None and len(post_vms) > 0:
+        logging.error("The scenario does not contain a service VM. Will not generate launch scripts for the post-launched VMs in the scenario.")
+        return 1
+    sos_vm_id = int(sos_vm_id)
+
+    try:
+        os.mkdir(out_dir)
+    except FileExistsError:
+        pass
+    except Exception as e:
+        logging.error(f"Cannot create output directory: {e}")
+        return 1
+
+    if user_vmid == 0:
+        post_vm_ids = [int(vm_scenario_etree.get("id")) - sos_vm_id for vm_scenario_etree in post_vms]
     else:
-        print("Launch files generate failed".format(output))
+        post_vm_ids = [user_vmid]
+
+    for post_vm_id in post_vm_ids:
+        vm_scenario_etree = eval_xpath(scenario_etree, f"//vm[@id = {sos_vm_id + post_vm_id}]")
+        vm_launch_etree = eval_xpath(launch_etree, f"//user_vm[@id='{post_vm_id}']")
+        if vm_scenario_etree is None:
+            logging.warning(f"Post-launched VM {post_vm_id} is not specified in the scenario XML and has no launch script generated.")
+            continue
+
+        if vm_launch_etree is None:
+            logging.warning(f"Post-launched VM {post_vm_id} is not specified in the launch XML and has no launch script generated.")
+            continue
+
+        script = generate_for_one_vm(board_etree, vm_scenario_etree, vm_launch_etree, post_vm_id)
+        script.write_to_file(os.path.join(out_dir, f"launch_user_vm_id{post_vm_id}.sh"))
 
-    return err_dic
+    return 0
 
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--board", help="the XML file summarizing characteristics of the target board")
+    parser.add_argument("--scenario", help="the XML file specifying the scenario to be set up")
+    parser.add_argument("--launch", help="the XML file specifying the parameters of post-launched VMs")
+    parser.add_argument("--user_vmid", type=int, default=0, help="the post-launched VM ID (as is specified in the launch XML) whose launch script is to be generated, or 0 if all post-launched VMs shall be processed")
```

Comment:
The success message writes out the VM name, so shouldn't we use a VM name here and not ID number?  @junjiemao1

Reply:
This parameter is mainly for QA and is thus defined in the same way as it was in its predecessor in order to minimize impact to the auto-test scripts. Eventually this script will not be invoked by the user directly, but rather via the configurator or via Makefile.